### PR TITLE
RavenDB-18298 Corax - Distinct

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18298

### Additional description

Implementation of Distinct operation for Corax. Also contains the optimization of querying.

### Type of change

- New feature
- Optimization

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

-No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 1]
```diff
@@ -55,66 +60,114 @@ public CoraxIndexReadOperation(Index index, Logger logger, Transaction readTrans
...
+                        skippedResults.Value++;
+                        markedAsSkipped = true;
+                    }
+
+                    return default;
+                }
+
+                if (isDistinctCount == false)
+                {
+                    Dictionary<string, Dictionary<string, string[]>> highlightings = null;
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Check the warning - `The variable 'highlightings' is assigned but its value is never used`

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 2]
```diff
@@ -55,66 +60,114 @@ public CoraxIndexReadOperation(Index index, Logger logger, Transaction readTrans
...
+                    if (query.Metadata.HasHighlightings)
+                    {
+                        throw new NotImplementedException($"{nameof(Corax)} doesn't support {nameof(Highlightings)} yet.");
+                    }
+
+                    ExplanationResult explanation = null;
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Same here

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 3]
```diff
@@ -380,13 +433,75 @@ public override IEnumerable<string> DynamicEntriesFields(HashSet<string> staticF
             }
         }
 
+        // private static void Skip2<TQueryMatch>(ref TQueryMatch result, int position, ref int read, Reference<int> skippedResults, out long fetchedResults, ref long[] ids,
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Should be removed?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 4]
```diff
@@ -380,13 +433,75 @@ public override IEnumerable<string> DynamicEntriesFields(HashSet<string> staticF
...
+
         private static void Skip<TQueryMatch>(ref TQueryMatch result, int position, ref int read, Reference<int> skippedResults, out long readCounter, ref long[] ids,
+            CoraxIndexQueryingScope queryingScope,
             CancellationToken token)
             where TQueryMatch : IQueryMatch
         {
             if (ids.Length is 0)
                 throw new OutOfMemoryException("Corax buffer has no memory.");
 
+            var retriever = default(RetrieverInput);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: and here

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 5]
```diff
@@ -395,6 +510,7 @@ public override IEnumerable<string> DynamicEntriesFields(HashSet<string> staticF
                 {
                     token.ThrowIfCancellationRequested();
                     read = result.Fill(ids);
+                    // queryingScope?.RecordAlreadyPagedItemsInPreviousPage(ids, token);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Please remove commented code

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Deleted

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Deleted

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs (hunk 3 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Yes

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs (hunk 4 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Leftover, deleted.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs (hunk 5 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Leftover, deleted.

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**